contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,847 @@
+    // Replicas, Selector, Template, VolumeClaimsTemplate, and ServiceName 
+    // omitted for brevity.
+	v1.PodTemplateSpec `json:"template"`
+
+	// TemplateGeneration is a monotonically increasing, 64 bit, integer used to 
+	// indicate the version of the of the PodTemplateSpec. If nil, the 
+	// StatefulSetController has not initialized its revision history,
+	// change tracking is not enabled, and all Pods will be created from 
+	// Template.
+	TemplateGeneration  *int64 `json:"templateGeneration"`
+
+	// GenerationPartition partitions the Pods in the StatefulSet by ordinal such 
+    // that all Pods with a lower ordinal will be created from the PodTemplate that
+    // represents the current revision of the StatefulSet's revision history and 
+    // all Pods with an a greater or equal ordinal will be created from the 
+    // PodTemplate that represents the target revision of the StatefulSet's 
+    // revision history.
+	GenerationPartition *int32 `json:"generationPartition,omitempty`
+
+	// RevisionHistoryLimit is the maximum number of PodTemplates that will 
+	// be maintained in the StatefulSet's revision history. The revision history
+	// consists of all revisions not reprented by a currently applied 
+	// PodTemplate. The default value is 2.
+	RevisionHisotryLimit int32 `json:revisionHistoryLimit,omitempty`
+}
+```
+
+The following modifications will be made to the StatefulSetStatus API object.
+
+```go
+ type StatefulSetStatus struct {
+    // ObservedGeneration and Replicas fields are omitted for brevity.
+
+ 	// CurrentTemplateGeneration, if not nil, is the generation of the 
+ 	// PodTemplate that was used to create Pods with ordinals in the sequence
+ 	// [0,CurrentReplicas).
+ 	CurrentTemplateGeneration  *int64 `json:"currnetTemplateGeneration,omitempty"`
+ 	
+ 	// UpdatedTemplateGeneration, if not nil, is the generation of the PodTemplate 
+ 	// that was used to create Pods with ordinals in the sequence 
+    // [Replicas - UpdatedReplicas, Replicas).
+ 	UpdatedTemplateGeneration *int64 `json:"updatedTemplateGeneration,omitempty"`
+ 	
+ 	// ReadyReplicas is the current number of Pods, created by the StatefulSet
+    // controller, that have a Status for Running and a Ready Condition.
+ 	ReadyReplicas int32 `json:"readyReplicas,omitempty"`
+ 	
+ 	// CurrentReplicas is the number of Pods created by the StatefulSet 
+    // controller from the PodTemplateSpec indicated by CurrentTemplateGeneration.
+ 	CurrentReplicas int32 `json:"currentReplicas,omitempty"`
+ 	
+ 	// UpdatedReplicas is the number of Pods created by the StatefulSet
+    // controller from the PodTemplateSpec indicated by TargetTemplateGeneration.
+ 	UpdatedReplicas int32 `json:"taretReplicas,omitempty"`
+}
+```
+
+Additionally, we introduce the following constant.
+
+```go 
+// StatefulSetTemplateGenerationLabel is the label applied to a PodTemplate or 
+// Pod to indicate the position of the object's Template in the revision 
+// history of a StatefulSet.
```

Comment:
it might be helpful to introduce, before this line, where you intend to keep the revision history